### PR TITLE
Add Parser support for Type System Extensions

### LINKED ISSUE #238: Support type system extensions in Parser
Schema extensions will make the codegen commands fail. We need to add support for it in the parser.

See http://spec.graphql.org/June2018/#TypeSystemExtension

### PR DESCRIPTION
**Pull request by jorge-vasquez-2301**
closes #238

### CODE REVIEW

File: core/src/main/scala/caliban/parsing/Parser.scala [hunk 1]
```diff
@@ -266,6 +268,62 @@ object Parser {
         )
     }
 
+  private def schemaExtension[_: P]: P[SchemaExtension] =
+    P("extend schema" ~/ directives.? ~ "{" ~ rootOperationTypeDefinition.rep ~ "}").map {
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/parsing/Parser.scala]: It should be possible to extend a schema and only provide directives without any operation: 
http://spec.graphql.org/June2018/#sec-Schema-Extension (2nd case)

File: core/src/main/scala/caliban/parsing/Parser.scala [hunk 2]
```diff
@@ -266,6 +268,62 @@ object Parser {
...
+    }
+
+  private def scalarTypeExtension[_: P]: P[ScalarTypeExtension] =
+    P("extend scalar" ~/ name ~ directives).map {
+      case (name, directives) =>
+        ScalarTypeExtension(name, directives)
+    }
+
+  private def objectTypeExtension[_: P]: P[ObjectTypeExtension] =
+    P("extend type" ~/ name ~ implements.? ~ directives.? ~ "{" ~ fieldDefinition.rep ~ "}").map {
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/parsing/Parser.scala]: Here too, there could be a case where we provide no fields and only implements/directives: http://spec.graphql.org/June2018/#sec-Object-Extensions

Same for the others below

**Comment by jorge-vasquez-2301** [comment]
@ghostdogpr Thanks for looking into this! And yeah, I totally missed the alternative cases, so I'll make sure to add them 😀

File: core/src/main/scala/caliban/parsing/Parser.scala [hunk 3]
```diff
@@ -266,6 +267,161 @@ object Parser {
...
+          Nil,
+          Nil,
+          fields.toList
+        )
+    }
+
+  private def objectTypeExtension[_: P]: P[ObjectTypeExtension] =
+    P(
+      "extend type" ~/ (
+        objectTypeExtensionWithInterfacesAndOptionalFields |
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/parsing/Parser.scala]: I think you can reduce to 3 cases:
- fields + optional interfaces + optional directives
- no fields + directives + optional interfaces
- no fields, no directives, interfaces

File: core/src/main/scala/caliban/parsing/Parser.scala [hunk 4]
```diff
@@ -266,6 +267,161 @@ object Parser {
...
+      SchemaExtension(
+        Nil,
+        opsMap.get(OperationType.Query).map(_.name),
+        opsMap.get(OperationType.Mutation).map(_.name),
+        opsMap.get(OperationType.Subscription).map(_.name)
+      )
+    }
+
+  private def schemaExtension[_: P]: P[SchemaExtension] =
+    P("extend schema" ~/ (schemaExtensionWithDirectivesAndOptionalOperations | schemaExtensionWithOperations))
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/parsing/Parser.scala]: How about sticking to the spec definitions and have:
- operations + optional directives
- no operations, directives

File: core/src/main/scala/caliban/parsing/Parser.scala (hunk 3 above)
Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]: @ghostdogpr Yeah, actually my first attempt was modelling the Parser more strictly after the specs, like this:
```scala
private def objectTypeExtensionWithOptionalInterfacesOptionalDirectivesAndFields[_: P]: P[ObjectTypeExtension] =
    P(name ~ implements.? ~ directives.? ~ "{" ~ fieldDefinition.rep ~ "}").map {
      case (name, implements, directives, fields) =>
        ObjectTypeExtension(
          name,
          implements.getOrElse(Nil),
          directives.getOrElse(Nil),
          fields.toList
        )
    }

  private def objectTypeExtensionWithOptionalInterfacesAndDirectives[_: P]: P[ObjectTypeExtension] =
    P(name ~ implements.? ~ directives).map {
      case (name, implements, directives) =>
        ObjectTypeExtension(
          name,
          implements.getOrElse(Nil),
          directives,
          Nil
        )
    }

  private def objectTypeExtensionWithInterfaces[_: P]: P[ObjectTypeExtension] =
    P(name ~ implements).map {
      case (name, implements) =>
        ObjectTypeExtension(
          name,
          implements,
          Nil,
          Nil
        )
    }


  private def objectTypeExtension[_: P]: P[ObjectTypeExtension] =
    P(
      "extend type" ~/ (
        objectTypeExtensionWithOptionalInterfacesOptionalDirectivesAndFields |
          objectTypeExtensionWithOptionalInterfacesAndDirectives |
          objectTypeExtensionWithInterfaces
      )
    )
```

Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]: However, that didn't work as expected, when running unit tests I've got failures for some cases. Here are the error messages I got (I've enabled verboseFailures temporally):

Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]:
```
 extend type with interfaces and directives
[info]     Fiber failed.
[info]     A checked error was not handled.
[info]     Parsing Error: Expected document:1:1 / objectTypeExtension:1:1 / objectTypeExtensionWithOptionalInterfacesOptionalDirectivesAndFields:1:13 / "{":1:58, found "" 
```

Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]:
```
extend type with directives
[info]     Fiber failed.
[info]     A checked error was not handled.
[info]     Parsing Error: Expected document:1:1 / objectTypeExtension:1:1 / objectTypeExtensionWithOptionalInterfacesOptionalDirectivesAndFields:1:13 / "{":1:33, found "" 
```

Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]: I've tried other combinations as well, but none of them worked for all cases, that's why I had to implement the parser rules in a different way.

Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]: What do you think @ghostdogpr? Maybe I'm missing something?

File: core/src/main/scala/caliban/parsing/Parser.scala (hunk 4 above)
Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]: Similarly to the above case, my first version was more like the spec definitions:

```scala
private def schemaExtensionWithOptionalDirectivesAndOperations[_: P]: P[SchemaExtension] =
    P(directives.? ~ "{" ~ rootOperationTypeDefinition.rep ~ "}").map {
      case (directives, ops) =>
        val opsMap = ops.toMap
        SchemaExtension(
          directives.getOrElse(Nil),
          opsMap.get(OperationType.Query).map(_.name),
          opsMap.get(OperationType.Mutation).map(_.name),
          opsMap.get(OperationType.Subscription).map(_.name)
        )
    }

  private def schemaExtensionWithDirectives[_: P]: P[SchemaExtension] =
    P(directives).map(SchemaExtension(_, None, None, None))

  private def schemaExtension[_: P]: P[SchemaExtension] =
    P("extend schema" ~/ (schemaExtensionWithOptionalDirectivesAndOperations | schemaExtensionWithDirectives))
```

File: core/src/main/scala/caliban/parsing/Parser.scala (hunk 3 above)
Comment (ghostdogpr) [reply core/src/main/scala/caliban/parsing/Parser.scala]: Hmm that’s strange indeed. I can’t see anything wrong with the code at first sight but tomorrow I’ll get your branch and try to figure it out in debug.

File: core/src/main/scala/caliban/parsing/Parser.scala (hunk 4 above)
Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]: However, this didn't work for all cases, I got an error for one of the unit tests:

```
extend schema with directives
[info]     Fiber failed.
[info]     A checked error was not handled.
[info]     Parsing Error: Expected document:1:1 / schemaExtension:1:1 / schemaExtensionWithOptionalDirectivesAndOperations:1:15 / "{":1:30, found "" 
```

Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]: I tried other combinations that were more similar to the specs too, but none of them worked

File: core/src/main/scala/caliban/parsing/Parser.scala (hunk 3 above)
Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]: And, I got similar issues when I tried to define parsers for the other type extension cases

Comment (jorge-vasquez-2301) [reply core/src/main/scala/caliban/parsing/Parser.scala]: Awesome, thanks for your help @ghostdogpr!

**Comment by ghostdogpr** [comment]
@jorge-vasquez-2301 OK, I found the reason. You just need to change the `~/` into `~` in `def directive`. Otherwise it takes the first path and "cuts" which means it won't go back and try the other possibilities.

**Comment by jorge-vasquez-2301** [comment]
Ohhhh, didn't realize that, thanks @ghostdogpr! I'll make the adjustments

**Pull request merged**